### PR TITLE
fix #61 and use "here documents"

### LINKED ISSUE #60: why using python venv inside container?
inside the docker files (standard and nightly) the project is setup as you would in a local dev-environment.

why? the container already gives that level of encapsulation we need.

am I missing a point here?

### LINKED ISSUE #61: why use wheels instead of venv?
followup to #60 

after investigating deeper into docker multistage I came across the topic of wheels vs venv.
we use both and I would like to understand why.

according to this article [https://pythonspeed.com/articles/multi-stage-docker-python/](https://pythonspeed.com/articles/multi-stage-docker-python/) a venv is perfectly capable of being passed from one stage to the other - as we do using the wheels directory.

this made me lift my eyebrow:

> "Another solution to the problem is to build all the packages as wheels—binary packages with the compiled code—and then copy the wheel files over to the build stage and install them there. This works, but doesn’t really add much."

a second source I found: [https://testdriven.io/blog/docker-best-practices/#using-python-virtual-environments](https://testdriven.io/blog/docker-best-practices/#using-python-virtual-environments) indicates even more concrete:
> "you may want to use a virtual environment in a multi-stage build rather than building wheel files."

is there any other benefit of using wheels over venv?

### PR DESCRIPTION
**Pull request by jensens**
This builds everything in stage 1 and then copy it. Keeps .cache in stage 1. Once we use layer caching in GHA (todo), this speeds up things.
Stage 2 is quite simple then.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -6,15 +6,17 @@ ENV PIP_VERSION=22.2.2
...
+    apt-get update
+    buildDeps="dpkg-dev gcc libbz2-dev libc6-dev libffi-dev libjpeg62-turbo-dev libldap2-dev libopenjp2-7-dev libpcre3-dev libpq-dev libsasl2-dev libssl-dev libtiff5-dev libxml2-dev libxslt1-dev wget zlib1g-dev python3-dev build-essential"
+    apt-get install -y --no-install-recommends $buildDeps
+    rm -rf /var/lib/apt/lists/* /usr/share/doc
+    python -m venv /venv
+    /venv/bin/pip install -U "pip==${PIP_VERSION}"
+    /venv/bin/pip install Plone plone.volto ${EXTRA_PACKAGES} -c https://dist.plone.org/release/$PLONE_VERSION/constraints.txt  ${PIP_PARAMS}
+    rm .cache -rf
+    find /venv \( -type f -a -name '*.pyc' -o -name '*.pyo' \) -exec rm -rf '{}' +
+EOT
```
Comment (davisagli) [inline Dockerfile:19]: @jensens We should double check that an error in any command still aborts the build when using the heredoc syntax

File: Dockerfile [hunk 2]
```diff
@@ -26,31 +28,23 @@ LABEL maintainer="Plone Community <dev@plone.org>" \
...
+    chown -R /data plone:plone
+    ln -s /data var
+EOT
 
 WORKDIR /app
-
-RUN python -m venv . \
-    && ./bin/pip install -U "pip==${PIP_VERSION}" \
-    && ./bin/pip install --force-reinstall --no-index --no-deps ${PIP_PARAMS} /wheelhouse/* \
-    && ./bin/pip install "relstorage==3.4.5" "psycopg2==2.9.3" ${PIP_PARAMS} \
```
Comment (davisagli) [inline Dockerfile:44]: We still need relstorage and psycopg2 to be installed somewhere, right?

(Also, maybe we can use psycopg2-binary?)

Comment (davisagli) [reply Dockerfile:44]: Oh, I just found them in EXTRA_PACKAGES

File: Dockerfile (hunk 1 above)
Comment (jensens) [reply Dockerfile:19]: it does :D - I just run into it

**Comment by jensens** [comment]
Should we introduce a changelog here? Mr.Roboto complains....

**Comment by davisagli** [comment]
@jensens I'm not sure about the changelog. It sounds like a good idea, but I'm not sure when and where it should get published if we use towncrier. If we don't want it, we can disable the check in mr.roboto for this repository.

This looks promising. I'd like to get @ericof's feedback as well.

How large is the image in the end, and how does it compare to the current one?

**Review by davisagli — APPROVED** [review]
Sounds nice! But, I don't understand: how does the workdir affect the size of the image?

**Comment by jensens** [comment]
fun with image sizes

TLDR:
We go down from 952MB to 519MB.

- no wheelhouse
- apt-clean

Also lesson learned: Set WORKDIR early (before it is populated) otherwise it adds the contents of the workdir on new layer. 

# Past

first, what we have before this PR:
```
~/ws/cdev/plone-backend (fix-60✔)$ docker pull plone/plone-backend:latest
latest: Pulling from plone/plone-backend
Digest: sha256:dc4f5b71db33ba7b1f3331de7d14d0cca23c15a0c1519e91bbba3d544678b234
Status: Image is up to date for plone/plone-backend:latest
docker.io/plone/plone-backend:latest
~/ws/cdev/plone-backend (fix-60✔)$ docker image ls plone/plone-backend:latest
REPOSITORY            TAG       IMAGE ID       CREATED        SIZE
plone/plone-backend   latest    3b8fc681dc9b   2 months ago   952MB
```
in detail:
```
$ docker history plone/plone-backend:latest
IMAGE          CREATED        CREATED BY                                      SIZE      COMMENT
3b8fc681dc9b   2 months ago   CMD ["start"]                                   0B        buildkit.dockerfile.v0
<missing>      2 months ago   ENTRYPOINT ["/app/docker-entrypoint.sh"]        0B        buildkit.dockerfile.v0
<missing>      2 months ago   HEALTHCHECK &{["CMD-SHELL" "wget -q http://1…   0B        buildkit.dockerfile.v0
<missing>      2 months ago   VOLUME [/data]                                  0B        buildkit.dockerfile.v0
<missing>      2 months ago   EXPOSE map[8080/tcp:{}]                         0B        buildkit.dockerfile.v0
<missing>      2 months ago   RUN /bin/sh -c ln -s /data var     && find /…   294MB     buildkit.dockerfile.v0
<missing>      2 months ago   COPY skeleton/ /app # buildkit                  13.5kB    buildkit.dockerfile.v0
<missing>      2 months ago   RUN /bin/sh -c python -m venv .     && ./bin…   298MB     buildkit.dockerfile.v0
<missing>      2 months ago   WORKDIR /app                                    0B        buildkit.dockerfile.v0
<missing>      2 months ago   RUN /bin/sh -c useradd --system -m -d /app -…   150MB     buildkit.dockerfile.v0
<missing>      2 months ago   COPY /wheelhouse /wheelhouse # buildkit         93.6MB    buildkit.dockerfile.v0
<missing>      2 months ago   LABEL maintainer=Plone Community <dev@plone.…   0B        buildkit.dockerfile.v0
<missing>      2 months ago   ENV PIP_VERSION=22.0.4                          0B        buildkit.dockerfile.v0
<missing>      2 months ago   ENV PIP_PARAMS=                                 0B        buildkit.dockerfile.v0
<missing>      3 months ago   /bin/sh -c #(nop)  CMD ["python3"]              0B        
<missing>      3 months ago   /bin/sh -c set -eux;   savedAptMark="$(apt-m…   11.3MB    
<missing>      3 months ago   /bin/sh -c #(nop)  ENV PYTHON_GET_PIP_SHA256…   0B        
<missing>      3 months ago   /bin/sh -c #(nop)  ENV PYTHON_GET_PIP_URL=ht…   0B        
<missing>      3 months ago   /bin/sh -c #(nop)  ENV PYTHON_SETUPTOOLS_VER…   0B        
<missing>      3 months ago   /bin/sh -c #(nop)  ENV PYTHON_PIP_VERSION=22…   0B        
<missing>      3 months ago   /bin/sh -c set -eux;  for src in idle3 pydoc…   0B        
<missing>      3 months ago   /bin/sh -c set -eux;   savedAptMark="$(apt-m…   29.1MB    
<missing>      3 months ago   /bin/sh -c #(nop)  ENV PYTHON_VERSION=3.8.13    0B        
<missing>      3 months ago   /bin/sh -c #(nop)  ENV GPG_KEY=E3FF2839C048B…   0B        
<missing>      3 months ago   /bin/sh -c set -eux;  apt-get update;  apt-g…   7.07MB    
<missing>      3 months ago   /bin/sh -c #(nop)  ENV LANG=C.UTF-8             0B        
<missing>      3 months ago   /bin/sh -c #(nop)  ENV PATH=/usr/local/bin:/…   0B        
<missing>      3 months ago   /bin/sh -c #(nop)  CMD ["bash"]                 0B        
<missing>      3 months ago   /bin/sh -c #(nop) ADD file:7f2320197e75c5169…   69.2MB  
```

# Future

With this PR we have 

```
~/ws/cdev/plone-backend (fix-60✔)$ docker image ls plone/plone-backend:6.0.0b3
REPOSITORY            TAG       IMAGE ID       CREATED         SIZE
plone/plone-backend   6.0.0b3   dba710ebe465   6 minutes ago   519MB
```

```
IMAGE          CREATED          CREATED BY                                      SIZE      COMMENT
dba710ebe465   4 minutes ago    CMD ["start"]                                   0B        buildkit.dockerfile.v0
<missing>      4 minutes ago    ENTRYPOINT ["/app/docker-entrypoint.sh"]        0B        buildkit.dockerfile.v0
<missing>      4 minutes ago    HEALTHCHECK &{["CMD-SHELL" "wget -q http://1…   0B        buildkit.dockerfile.v0
<missing>      4 minutes ago    VOLUME [/data]                                  0B        buildkit.dockerfile.v0
<missing>      4 minutes ago    EXPOSE map[8080/tcp:{}]                         0B        buildkit.dockerfile.v0
<missing>      4 minutes ago    RUN /bin/sh -c     useradd --system -m -d /a…   150MB     buildkit.dockerfile.v0
<missing>      15 minutes ago   COPY /app /app # buildkit                       244MB     buildkit.dockerfile.v0
<missing>      15 minutes ago   WORKDIR /app                                    0B        buildkit.dockerfile.v0
<missing>      15 minutes ago   LABEL maintainer=Plone Community <dev@plone.…   0B        buildkit.dockerfile.v0
<missing>      3 days ago       /bin/sh -c #(nop)  CMD ["python3"]              0B        
<missing>      3 days ago       /bin/sh -c set -eux;   savedAptMark="$(apt-m…   11.4MB    
<missing>      3 days ago       /bin/sh -c #(nop)  ENV PYTHON_GET_PIP_SHA256…   0B        
<missing>      3 days ago       /bin/sh -c #(nop)  ENV PYTHON_GET_PIP_URL=ht…   0B        
<missing>      3 days ago       /bin/sh -c #(nop)  ENV PYTHON_SETUPTOOLS_VER…   0B        
<missing>      3 days ago       /bin/sh -c #(nop)  ENV PYTHON_PIP_VERSION=22…   0B        
<missing>      3 days ago       /bin/sh -c set -eux;  for src in idle3 pydoc…   0B        
<missing>      3 days ago       /bin/sh -c set -eux;   savedAptMark="$(apt-m…   29.8MB    
<missing>      3 days ago       /bin/sh -c #(nop)  ENV PYTHON_VERSION=3.9.15    0B        
<missing>      11 days ago      /bin/sh -c #(nop)  ENV GPG_KEY=E3FF2839C048B…   0B        
<missing>      11 days ago      /bin/sh -c set -eux;  apt-get update;  apt-g…   3.1MB     
<missing>      11 days ago      /bin/sh -c #(nop)  ENV LANG=C.UTF-8             0B        
<missing>      11 days ago      /bin/sh -c #(nop)  ENV PATH=/usr/local/bin:/…   0B        
<missing>      12 days ago      /bin/sh -c #(nop)  CMD ["bash"]                 0B        
<missing>      12 days ago      /bin/sh -c #(nop) ADD file:b78b777208be08edd…   80.5MB  
```

**Comment by jensens** [comment]
> Sounds nice! But, I don't understand: how does the workdir affect the size of the image?

**Comment by jensens** [comment]
> Sounds nice! But, I don't understand: how does the workdir affect the size of the image?

No idea. I just found it (read from bottom to top):

```
<missing>      15 hours ago   WORKDIR /app                                    244MB     buildkit.dockerfile.v0
<missing>      15 hours ago   RUN /bin/sh -c     useradd --system -m -d /a…   394MB     buildkit.dockerfile.v0
<missing>      15 hours ago   COPY /app /app # buildkit                       244MB     buildkit.dockerfile.v0
```

versa

```
<missing>      11 seconds ago   RUN /bin/sh -c     useradd --system -m -d /a…   394MB     buildkit.dockerfile.v0
<missing>      22 seconds ago   COPY /app /app # buildkit                       244MB     buildkit.dockerfile.v0
<missing>      28 seconds ago   WORKDIR /app                                    0B        buildkit.dockerfile.v0
```
